Add description, owner, and lifecycle to the catalog model

freben opened the pull request:
See #983

This does NOT limit the set of valid strings that form a lifecycle. We can add a validation later for that but I am not sure if we should limit things at that level. In any case it should be flexible enough to be extended by organizations.

One way to do that would be to add an `isValidLifecycle` to [the overridable validators](https://github.com/spotify/backstage/blob/master/packages/catalog-model/src/validation/types.ts#L17), but that's left open still.

stefanalund commented on plugins/catalog-backend/fixtures/one_component.yaml:
```diff
@@ -4,3 +4,5 @@ metadata:
   name: component3
```
Can this be updated to something humanly understandable as well?

stefanalund replied:
Also, why not add `description`?

freben replied:
Well I mainly wanted to unbreak them. These will all soon be thrown away entirely, and be replaced with an `example-entities.yaml` file instead.

freben commented:
@stefananlund Note also that this sets both lifecycle and owner as required. Is that sensible?

stefanalund commented:
Yes. Other adopting companies may have different opinions/rules, but I think these are sensible.

freben commented:
@stefanalund But the PR makes them fully, for everybody everywhere, required. At least the way that things work today.

Rugvip commented:
Beginning to think we should take a documentation-first approach to the fields. If they docs don't make sense and we can't describe how every company that adopts Backstage is supposed to use the field, we're probably on the wrong track.

Rugvip commented on packages/catalog-model/src/entity/policies/ReservedFieldsEntityPolicy.ts:
```diff
@@ -17,47 +17,70 @@
+type ExceptWhere = 'metadata' | 'spec' | 'labels' | 'annotations';
+type ReservedFields = Record<string, ExceptWhere[]>;
+
+const DEFAULT_RESERVED_ENTITY_FIELDS: ReservedFields = {
```
Feel it would be nice to have this constant be a list of `.` separated names, i.e. `['apiVersion', 'kind', 'spec', 'metadata.uid']` etc., and then convert to whatever format is best for the runtime check.

freben replied:
done

freben commented:
@stefanalund @Rugvip OK, added a doc (where #1174 suggests it should be) with as much detail as I could muster. If this is accepted we can strip down the ADR to mainly the overall decision, and let the actual in depth documentation of the format be in this new file. I also rewrote `ReservedFieldsEntityPolicy` as requested.

hooloovooo commented on docs/features/software-catalog/descriptor-format.md:
```diff
@@ -0,0 +1,295 @@
+- Strip out all characters outside of the set `[a-zA-Z0-9]`
+- Convert to lowercase
+
+Example: `CircleciBuildsDs_avro_gcs` -> `circlecibuildsdsavrogcs`
```
Wouldn't this be CircleciBuildsDsavrogcs if all but `[a-zA-Z0-9]` is stripped? (Genuine question since my regex-fu is bad.

freben replied:
Note that it also says "convert to lowercase" after that :)

hooloovooo replied:
Ohh, sorry. It all checks out then. :)

stefanalund commented on docs/features/software-catalog/descriptor-format.md:
```diff
@@ -0,0 +1,295 @@
+# Descriptor Format of Catalog Entities
```
❤️  for adding docs!

jcarres-mdsol commented on docs/features/software-catalog/descriptor-format.md:
```diff
@@ -0,0 +1,295 @@
+- `deprecated` - a component that is at the end of its lifecycle, and may disappear at a
+  later point in time
+
+### `spec.owner` [required]
```
I'm concerned that `owner` has different meanings for different people. Depending on the organization's internal structure, maybe they have an architecture team, for them owner is the architect, but for managers it is the engineering manager. As you said, probably team is the right option here usually, unless the component is for instance a huge service with 5 teams working on it. Or maybe that organization does not use email or any obvious way to address the team.

jcarres-mdsol commented on docs/features/software-catalog/descriptor-format.md:
```diff
@@ -0,0 +1,295 @@
+The software catalog accepts any lifecycle value, but an organisation should take great
+care to establish a proper taxonomy for these.
+
+The current set of well-known and common values for this field is:
```
What does this mean? It means that other values are not allowed?
FWIW, we were thinking of adding this (inside metadata) and our values also included `pre_production` for all those things that passed experimental, probably they are in some sandbox or develop environment somewhere but not in production yet.

freben replied:
Sounds good!

It's surprisingly hard to nail the semantics of flags like these. They can mean different things from different perspectives/organisations, and based on one's previous experiences and preferences.

When talking about preprod, what can one read into that?

- That it's "physically" in the preprod (staging like) environment - as in, it speaks of deployment states?
- That it has reached that certain step before publicly-reachable production in the business processes of the org - as in, it speaks of ways of working?
- That it is considered not-yet-fully-stable but can be consumed if you are willing to accept the consequences - as in, it speaks of maturity or SLAs?

Or even worse - a combination of the above?

This is not me bashing on the choice of `pre_production` at all, because I am certain it makes perfect sense within your context. But I think it can be interesting to mention, and this is the reason that we need to put up a few well known values like these and be very clear about what their intended usage is, while also letting people use other values as they see fit.

Overall - thanks for your feedback :)

jcarres-mdsol commented on docs/features/software-catalog/descriptor-format.md:
```diff
@@ -0,0 +1,295 @@
+type component may present tooling in the Backstage interface that is specific to just
+websites.
+
+The current set of well-known and common values for this field is:
```
We were also thinking about adding this, our values are (copy/paste from our docs): 
- web_frontend - Indicates that the Project is a Web Application’s Frontend Service
- web_backend - Indicates that the Project is a Web Application’s Backend Service
- web_fullstack - Indicates that the Project contains a Web Application’s Full Stack of Services including both Frontend & Backend Services
- mobile - Indicates that the Project is a Mobile Application Frontend App

`web_backend` == `service` but we did this for consistency with the other values.
Consider all this, hopefully you find it is a good idea 👯

freben replied:
In general when we go around our company to talk to users when trying to establish a new `kind` or `type`, we find that it's not easy to find values that everybody agrees on. Particularly since the inevitable question becomes, WHY do you want to discern type A from type B? Who consumes the value? What is the actual effect/outcome of switching from one value to the other?

We are pretty coarsely granular in our choices. We tend to choose types that are tied to how humans want to consume the thing and what broad family of behaviors they represent. So we have `service`, which is any kind of deployable service-like thing that is communicated with via APIs, without splitting it into subtypes. The reason is, that the greatest consumer of the field is Backstage itself - it may choose to enable service-specific tooling and subviews when looking at that entity's page. The same tools may not appear when looking at a library type entity, because it isn't ever conceivably deployed to kubernetes for example.

You may want to have finer granularity and that could be fine too. But if you find that they actually BEHAVE the exact same, for example in terms of infra tooling that applies to them, then maybe their difference is better explained in the README of the component or in the description field, than as specific types.

Your web_fullstack service may even be represented as linked to three APIs - one that's serving the frontend, one that's an internal management interface, and one that's the graphql backend part. Maybe a Backstage user can then see at a glance what it does anyway, if the entity view is well made, without subtyping.

Again, your choices, and everybody's needs won't be the same as ours. Therefore it's left open to extension :)

JordiPolo replied:
It is hard to tell if applies without seeing it in more details.
I'm concerned about if this field MUST have these values or not.
It sounds like backstage will know `service` for instance. If so, then these need to be pretty well defined and making sure covers lots of usecases. Website for instance not sure what is that, maybe anything rendering html. 
Or alternatively Backstage does not know about this, it will render them as it finds them in the DB and will match plugins when those target them but really do not know more than that.

jcarres-mdsol commented on docs/features/software-catalog/descriptor-format.md:
```diff
@@ -0,0 +1,295 @@
+
+### `metadata` [required]
+
+A structure that contains metadata about the entity, i.e. things that aren't directly
```
I think it is becoming increasingly difficult to know the difference between metadata and spec.
In this PR, why `lifecycle` is spec and not metadata? 
Maybe things should be divided in some other way or not divided at all and all be inside `spec`?

JordiPolo replied:
it is indeed weird that this thing is so important to be here, same for lifecycle.
As I said, I do not really have an answer for "owner" that in general makes sense so I'll just put there the name of the company.

stefanalund commented on docs/features/software-catalog/descriptor-format.md:
```diff
@@ -0,0 +1,295 @@
+### `apiVersion` and `kind` [required]
+
+The `kind` is the high level entity type being described. The
+[Backstage system model](https://github.com/spotify/backstage/issues/390) describes
```
This issue will be closed at some point. I would prefer if you linked to a permanent location for this information, an ADR for example.

Also, should we not specify examples of `Kind`s that are supported now? Maybe link to the paragraph further down.

freben replied:
@stefanalund OK, pointed to ADR005 instead and linked to the Component section.

stefanalund commented:
Let's change versioning to `alpha` and then merge this one.

hooloovooo commented:
Is there a reason spec is not defined as a type other than object?

freben commented:
In theory we could allow other types, but I think it would be poor form. The semantics of that value would be too unclear in my opinion. It will always be better to add a `{ someFieldNameWithSemantics: <thevalue> }` rather than just `<thevalue>` in terms of conveying what it's for, and in terms of extensibility.

Since every single kind/type I've seen so far has used an object form here, it felt that making it more flexible than that would just be adding risk, rather than benefit.
